docs/source/plugin/what.rst

```diff
@@ -0,0 +1,154 @@
+
+* everything that is not available in this ``__init__.py`` file won't be
+  visible to Sopel
+* either:
+   * dependencies should be limited to Sopel and Python's built-in library,
+     for ease of installation; or
+   * the plugin is a "one-off", not intended to be shared
+
+By default, Sopel's plugin directories are **not** available in ``sys.path``,
+so it is not possible to do ``from my_plugin_dir import some_module`` without
+a specific configuration from the bot's owner.
+
+When a plugin author wants to split code into different files, and/or share it
+between multiple plugins, the best option is to use an :ref:`Entry point`
+plugin instead.
+
+.. note::
+
+   Just like single file plugins, you *can* use extra dependencies with folder
+   plugins; the same advice about ``requirements.txt`` and ``pip`` applies.
+
+Namespace package
+-----------------
+
+A **Namespace package** plugin is a Python :term:`namespace package`, using
+the namespace ``sopel_modules``. It must be installed in the Python environment
+to be used by Sopel, and can require Python dependencies.
+
+Given a ``sopel_modules.plugin`` plugin, Sopel will load everything that is
+available from the ``sopel_modules/plugin/__init__.py`` file.
+
+It is the initial version of Sopel's packaged plugins: it can be packaged and
+uploaded to `PyPI`_ then installed using ``pip install``.
+
+When a plugin author wants to distribute a Sopel plugin, the best option is to
+use an :ref:`Entry point` plugin instead.
+
+Entry point
+-----------
+
+.. versionadded:: 7.0
+
+An **Entry point** plugin is a Python module or package distributed via a
+``setup.py`` script, and it is available to Sopel via Sopel's ``sopel.plugins``
+`setup entry point`__.
+
+Given this definition of an entry point from a ``setup.cfg`` file::
+
+   [options.entry_points]
+   sopel.plugins =
+       my_plugin = package_name.my_plugin_file
+
+Sopel will load everything available from the ``package_name.my_plugin_file``
+Python module under the plugin name ``my_plugin``. It means that you can have
+any package name and any module name as long as it is a valid Python module
+and as long as you properly define the entry point.
+
+You should write an entry point plugin when:
+
+* you want to distribute your plugin on `PyPI`_
+* you want to split the code in multiple files
+* you have dependencies beyond Sopel and Python's standard library
+* you want a modern and reliable way to package your Sopel plugin
+* you want to distribute more than one Sopel plugin per distributed package
+
+An entry point plugin is the best, most flexible way to package and distribute
+a Sopel plugin (or collection of plugins) in a clean, easy-to-update manner.
+
+Note that a single Python distributed package can expose more than one Sopel
+entry point plugin, which is great to bundle multiple plugins at once.
+
+.. seealso::
+
+   The PyPA specification explains in its `Entry points specification`_ what
+   entry points are and how to use them.
```

Comment:
```suggestion
   The Python Packaging Authority explains how entry points work and how to
   use them in its `Entry points specification`_.
```

I'm mostly avoiding the use of "specification" twice, but expanding the acronym for new Python users who might not know what PyPA is. (Feel free to reject that idea and use the suggestion below instead, which only eliminates the redundant word.)

```suggestion
   The PyPA explains in its `Entry points specification`_ what entry points
   are and how to use them.
```